Index: abiword-20121011/src/Makefile.am
===================================================================
--- abiword-20121011.orig/src/Makefile.am
+++ abiword-20121011/src/Makefile.am
@@ -163,8 +163,11 @@ if HAVE_INTROSPECTION
 
 INTROSPECTION_GIRS += Abi-@ABIWORD_SERIES@.gir
 
-Abi_@ABIWORD_SERIES@.gir: $(INTROSPECTION_SCANNER) libabiword-@ABIWORD_SERIES@.la Makefile
-Abi_@ABI_SUB_SERIES@_gir_INCLUDES = GObject-2.0 Gtk-3.0 Gsf-1
+Abi-@ABIWORD_SERIES@.gir: $(INTROSPECTION_SCANNER) libabiword-@ABIWORD_SERIES@.la Makefile
+# FIXME Gsf-1 introspection bindings fail to create, so remove it from
+# includes for now.
+Abi_@ABI_SUB_SERIES@_gir_INCLUDES = GObject-2.0 Gtk-3.0
+Abi_@ABI_SUB_SERIES@_gir_PACKAGES = libgsf-1
 Abi_@ABI_SUB_SERIES@_gir_CFLAGS = $(libabiword_@ABIWORD_SERIES@_la_CPPFLAGS)
 Abi_@ABI_SUB_SERIES@_gir_LIBS = libabiword-@ABIWORD_SERIES@.la
 Abi_@ABI_SUB_SERIES@_gir_FILES = \
